plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java

```diff
@@ -399,36 +393,23 @@ public String getName() {
   }
 
   private void updatePipelineOptionsForm() {
-    final SettableFuture<Map<PipelineOptionsType, Set<PipelineOptionsProperty>>> optionsHierarchyFuture =
-        SettableFuture.create();
-    optionsHierarchyFuture.addListener(new Runnable() {
-      @Override
-      public void run() {
-        if (internalComposite.isDisposed()) {
-          return;
-        }
-        BusyIndicator.showWhile(internalComposite.getDisplay(), new Runnable() {
-          @Override
-          public void run() {
-            try {
-              pipelineOptionsForm.updateForm(launchConfiguration, optionsHierarchyFuture.get());
-              updateLaunchConfigurationDialog();
-            } catch (InterruptedException | ExecutionException ex) {
-              DataflowUiPlugin.logError(ex, "Exception while updating available Pipeline Options"); //$NON-NLS-1$
-            }
-          }
-        });
-      }
-    }, displayExecutor);
     try {
+      // TODO: use "Holder" instead of "AtomicReference" when Java 8 is available.
+      final AtomicReference<Map<PipelineOptionsType, Set<PipelineOptionsProperty>>>
+          optionsHierarchy = new AtomicReference<>();
       // blocking call (regardless of "fork"), returning only after the inner runnable completes
       getLaunchConfigurationDialog().run(true, true, new IRunnableWithProgress() {
         @Override
         public void run(IProgressMonitor monitor)
             throws InvocationTargetException, InterruptedException {
-          optionsHierarchyFuture.set(launchConfiguration.getOptionsHierarchy(hierarchy));
+          monitor.beginTask(Messages.getString("updating.pipeline.options"), //$NON-NLS-1$
+              IProgressMonitor.UNKNOWN);
+          optionsHierarchy.set(launchConfiguration.getOptionsHierarchy(hierarchy));
         }
       });
+
+      pipelineOptionsForm.updateForm(launchConfiguration, optionsHierarchy.get());
```

Comment:
This `updateForm` was expensive and hence the `BusyIndicator`.

Reply:
I removed it after noticing that it doesn't work on my machine, but maybe it's only my machine? I think I rather put this into `run()`, which does work on my machine.

Reply:
Hmm... putting it `run()` is worse, as it consumes 100% of UI operations and the progress bar is frozen.

Reply:
Well, actually putting it into `run()` seems better. The 100% UI consumption seems inevitable anyway. (Basically, there is nothing much you can do if drawing UI itself is taking time.)

Reply:
You can't put it into run() as you're forking the job (i.e., it's in a non-UI thread).  Could you add some variable comments to the run (i.e., `true /*fork*/`)?

Reply:
I used `syncExec()`. I'll add comments to variables.